lib/azure/armrest/configuration.rb

```diff
@@ -0,0 +1,122 @@
+module Azure
+  module Armrest
+    class Configuration
+      # Used to store unique token information.
+      @@tokens = Hash.new([])
+
+      # The client ID used to gather token information.
+      attr_accessor :client_id
+
+      # The client key used to gather token information.
+      attr_accessor :client_key
+
+      # The tenant ID used to gather token information.
+      attr_accessor :tenant_id
+
+      # The subscription ID used in requests and to gather token information.
+      attr_accessor :subscription_id
+
+      # The resource group used for a given http request.
+      attr_accessor :resource_group
+
+      # The API version used for a given http request. The default is 2015-01-01.
+      attr_accessor :api_version
+
+      # The grant type. The default is client_credentials
+      attr_accessor :grant_type
+
+      # The content type specified for http requests. The default is 'application/json'
+      attr_accessor :content_type
+
+      # The accept type specified for http request results. The default is 'application/json'
+      attr_accessor :accept
+
+      # Set the time in which the token expires.
+      attr_writer :token_expiration
+
+      # Explicitly set the token.
+      attr_writer :token
+
+      # Yields a new Azure::Armrest::Configuration objects. Note that you must
+      # specify a client_id and client_key. All other parameters are optional.
+      #
+      # Example:
+      #
+      #   config = Azure::Armrest::Configuration.new(
+      #     :client_id  => 'xxxx',
+      #     :client_key => 'yyyy',
+      #     :tenant_id  => 'zzzz'
+      #   )
+      #
+      def initialize(hash)
+        hash.each{ |key,value| send("#{key}=", value) }
+
+        unless client_id && client_key
+          raise ArgumentError, "client_id and client_key must both be specified"
+        end
+
+        # Set defaults if not already set
+        @accept       ||= 'application/json'
+        @content_type ||= 'application/json'
+        @grant_type   ||= 'client_credentials'
+        @api_version  ||= '2015-01-01'
+      end
+
+      # A combination of grant_type, tenant_id, client_id and client_key,
+      # as a single string joined by underscores.
+      #
+      # Used to identify unique sessions.
+      #
+      def as_cache_key
+        "#{grant_type}_#{tenant_id}_#{client_id}_#{client_key}"
+      end
+
+      alias cache_key as_cache_key
+
+      # Returns the token for the given cache_key, or sets it if it does not
+      # exist or it has expired.
+      #
+      def token
+        @token, @token_expiration = @@tokens[cache_key] if @token.nil?
+
+        if @token.nil? || Time.now > (@token_expiration || Time.new(0))
```

Comment:
Since expiration can be set from outside, we'd better leave it as type `Date`. However the comparison here should change to `Time.now.utc > (@token_expiration.utc || Time.new(0))`. 

Other improvements can be to initialize `@token_expiration` to `Time.new(0)` in the constructor, and in the setter of `@token_expiration` to convert the input date to correct zone.